plugins/3MFWriter/ThreeMFWriter.py

```diff
@@ -253,7 +257,64 @@ def _storeMetadataJson(metadata: Dict[str, List[Dict[str, str]]], archive: zipfi
         metadata_file = zipfile.ZipInfo(path)
         # We have to set the compress type of each file as well (it doesn't keep the type of the entire archive)
         metadata_file.compress_type = zipfile.ZIP_DEFLATED
-        archive.writestr(metadata_file, json.dumps(metadata, separators=(", ", ": "), indent=4, skipkeys=True, ensure_ascii=False))
+        archive.writestr(metadata_file,
+                         json.dumps(metadata, separators=(", ", ": "), indent=4, skipkeys=True, ensure_ascii=False))
+
+    @staticmethod
+    def _getPluginPackageMetadata() -> List[Dict[str, str]]:
+        """Get metadata for all backend plugins that are used in the project.
+
+        :return: List of material metadata dictionaries.
+        """
+
+        backend_plugin_enum_value_regex = re.compile(
+            r"PLUGIN::(?P<plugin_id>\w+)@(?P<version>\d+.\d+.\d+)::(?P<value>\w+)")
+        # This regex parses enum values to find if they contain custom
+        # backend engine values. These custom enum values are in the format
+        #      PLUGIN::<plugin_id>@<version>::<value>
+        # where
+        #  - plugin_id is the id of the plugin
+        #  - version is in the semver format
+        #  - value is the value of the enum
+
+        plugin_ids = set()
+
+        def add_plugin_ids_in_stack(stack: CuraContainerStack):
+            for key in stack.getAllKeys():
+                value = str(stack.getProperty(key, "value"))
+                for plugin_id, _version, _value in backend_plugin_enum_value_regex.findall(value):
+                    plugin_ids.add(plugin_id)
+
+        # go through all stacks and find all the plugin id contained in the project
+        global_stack = Application.getInstance().getMachineManager().activeMachine
+        add_plugin_ids_in_stack(global_stack)
+
+        for container in global_stack.getContainers():
+            add_plugin_ids_in_stack(container)
+
+        for extruder_stack in global_stack.extruderList:
+            add_plugin_ids_in_stack(extruder_stack)
+
+            for container in extruder_stack.getContainers():
+                add_plugin_ids_in_stack(container)
+
```

Comment:
```suggestion
        for stack in chain(global_stack, global_stack.extruderList):
                add_plugin_ids_in_stack(stack)

                for container in stack.getContainers():
                        add_plugin_ids_in_stack(container)
```

Less code duplication.